jax/experimental/export/_shape_poly.py

```diff
@@ -587,13 +534,57 @@ def get_vars(self) -> set[str]:
       acc.update(mon.get_vars())
     return acc
 
+  @classmethod
+  def _merge_sorted_terms(
+      cls,
+      e1: Sequence[tuple[_DimMon, int]], i1: int, f1: int,
+      e2: Sequence[tuple[_DimMon, int]], i2: int, f2: int) -> Sequence[tuple[_DimMon, int]]:
+    """Computes e1[i1:] * f1 + e2[i2:] * f2.
+
+    e1, e2, and the result are sorted with largest term first.
+    This is an optimization for a common operation. The unoptimized code would
+    compute each subexpression in term.
+    """
+    #
+    acc = []
+    while True:
```

Comment:
This looks okay, but I thought I'd still offer a slightly less indented alternative:

```
while i1 < len(e1) and i2 < len(e2):
  ...

while i1 < len(e1):  # e2 is finished
  ...  

while i2 < len(e2):  # e1 is finished
  ...
```

Note that the last two `while` loops can also be generator expressions:

```
acc.extend((m2, m2_c * f2) for m2, m2_c in e2[i2:])
```

at the cost of copying the slice of `e2`, or even

```
import itertools as it
# ...
acc.extend((m2, m2_c * f2) for m2, m2_c in it.islice(e2, i2))
```

with no copying, but some redundant iteration over `e2`.

Reply:
Done. Very nice!